# Nigeria Food Price Inflation — What It Means for Your Pocket
### WFP market price monitoring, 2002 – 2026

Each section below is **one analysis in one code cell**, followed by a plain-language box
explaining what the result means for an ordinary household ("the man in the street").

| | |
|---|---|
| **Data** | World Food Programme (WFP) market price survey — `wfp_food_prices_nga.csv` |
| **Tools** | DuckDB (SQL), pandas, scikit-learn, Plotly |
| **Run** | *Kernel → Restart & Run All*. Cells must run in order; every cell only depends on cell 1 and 2. |

## What was wrong with the previous version of this analysis

The earlier notebook produced many attractive charts, but most of the numbers were not valid.
The main problems, in order of seriousness:

1. **Different pack sizes were averaged together.** The data mixes *100 kg wholesale sacks*
   (₦10,000 – ₦50,000) with *250 g retail sachets* (₦160). Every chart that used `AVG(price)`
   — the national price index, YoY inflation, state averages, seasonal pattern, category trend,
   box plots, and the "Maize/Millet/Sorghum" forecasts — was measuring *which pack sizes happened
   to be surveyed that month*, not inflation. Example: "average Millet price" collapses in 2023 only
   because the 100 kg sack series stopped being collected.
2. **The forecast back-test compared predictions with the wrong months.** In the AR model the
   test predictions were for Mar 2024 → Feb 2026, but they were scored against actual prices from
   Feb 2023 → Jan 2025 (13-month misalignment). All MAPE / skill numbers were therefore meaningless.
3. **The survey footprint changed, and this was ignored.** From February 2023 only Borno, Yobe and
   Adamawa (North-East) are monitored. "National" results after 2023 describe the North-East only.
   The header also claimed 118 markets; the file contains 68.
4. **Broken and duplicated cells.** "Kernel 7" was pasted into the middle of a tuning cell, which
   crashed (`food_prices does not exist`, because Kernel 6 replaced the database connection).
   Other cells used variables that never existed (`df`, `lstm_forecast`, `X_tr_s`, `ms`,
   `month_names`, `latest_ews`). The "LSTM" section contained a second copy of the AR model.
   The methods list promised ARIMA and LSTM, but neither was used.
5. **Wrong statistics.** Correlations were computed on price *levels* (any two rising prices look
   correlated); the "top 15" query actually took 25; the early-warning "shock" label (any item
   > 25 % YoY) was true almost all the time in a 30 %-inflation economy; the insight text under the
   charts was hard-coded rather than computed.
6. **Files.** 22 HTML chart exports (~110 MB, each embedding a full copy of Plotly), stale
   `.ipynb_checkpoints`, an empty `prices.duckdb`, an empty `requirements.txt`, and a stray editor
   worktree were cluttering the repository. These have been removed.

**The fix used throughout this notebook:** prices are only ever compared **with themselves** —
the same product, same pack size, same market, one month apart. These month-to-month changes
are then chained into an index. This is the same "matched sample" idea statistics offices use
for the Consumer Price Index.

## 1. The data — what we actually have

In [1]:
# ================================================================
# 1. SETUP, LOAD & DATA-COVERAGE CHECK
# ================================================================
import re
from pathlib import Path

import duckdb
import numpy as np
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

pd.set_option("display.width", 160)
TEMPLATE  = "plotly_white"
SAVE_HTML = False              # True → also export every chart to ./charts/*.html
CHART_DIR = Path("charts")


def show(fig, name, how_to_read, legend_title=""):
    """Display a chart with a legend on the right and a plain-English reading guide underneath
    (and optionally save a small CDN-linked HTML copy)."""
    note_lines = how_to_read.count("<br>") + 1
    bottom = 90 + 18 * note_lines                           # room for the x-axis + the note
    height = (fig.layout.height or 450) + 18 * note_lines
    top = fig.layout.margin.t if fig.layout.margin.t is not None else 110
    plot_h = height - top - bottom
    fig.update_layout(
        template=TEMPLATE, height=height, showlegend=True,
        margin=dict(t=top, b=bottom),
        legend=dict(title_text=f"<b>{legend_title}</b>" if legend_title else "",
                    orientation="v", x=1.02, xanchor="left", y=1, yanchor="top",
                    bgcolor="rgba(255,255,255,0.85)", bordercolor="#ccc", borderwidth=1))
    fig.add_annotation(
        text="<b>How to read this chart:</b> " + how_to_read,
        xref="paper", yref="paper", x=0, y=-70 / plot_h, xanchor="left", yanchor="top",
        showarrow=False, align="left", font=dict(size=12, color="#37474F"),
        bgcolor="#F5F7FA", bordercolor="#CFD8DC", borderwidth=1, borderpad=6)
    if SAVE_HTML:
        CHART_DIR.mkdir(exist_ok=True)
        fig.write_html(CHART_DIR / f"{name}.html", include_plotlyjs="cdn")
    fig.show()


def legend_item(fig, name, color, line=False, dash=None, legendgroup=None, **subplot):
    """Legend-only entry, for things that have no trace of their own (shaded bands, event lines)."""
    if line:
        style = dict(mode="lines", line=dict(color=color, dash=dash, width=2))
    else:
        style = dict(mode="markers", marker=dict(color=color, symbol="square", size=14))
    fig.add_trace(go.Scatter(x=[None], y=[None], name=name, hoverinfo="skip", legendgroup=legendgroup, **style),
                  **subplot)


def unit_size(unit):
    """'100 KG' → 100, '400 G' → 0.4, 'L' → 1, '30 pcs' → 30."""
    qty, name = re.match(r"^\s*([\d.]+)?\s*(\w+)", unit).groups()
    qty = float(qty) if qty else 1.0
    return qty / 1000 if name.upper() == "G" else qty


def base_unit(unit):
    name = unit.split()[-1].upper()
    return {"KG": "kg", "G": "kg", "L": "litre", "PCS": "piece",
            "TUBERS": "tuber", "UNIT": "unit"}[name]


LOCAL_CSV = Path("wfp_food_prices_nga.csv")
SOURCE = (LOCAL_CSV if LOCAL_CSV.exists() else
          "https://raw.githubusercontent.com/SmartDvi/food_inflation/refs/heads/main/wfp_food_prices_nga.csv")

con = duckdb.connect()
con.execute(f"CREATE OR REPLACE TABLE raw AS SELECT * FROM read_csv_auto('{SOURCE}')")

units = con.execute("SELECT DISTINCT unit FROM raw").fetchdf()
units["unit_size"] = units["unit"].map(unit_size)
units["base_unit"] = units["unit"].map(base_unit)
con.register("units_df", units)

# One clean table used by every analysis below.
# product = commodity + pack size + price type → the thing whose price we track over time.
con.execute("""
    CREATE OR REPLACE TABLE prices AS
    SELECT r.date::DATE                              AS date,
           DATE_TRUNC('month', r.date)::DATE         AS month,
           YEAR(r.date)                              AS year,
           r.admin1 AS state, r.market, r.category, r.commodity, r.unit, r.pricetype,
           r.commodity || ' [' || r.unit || ', ' || r.pricetype || ']' AS product,
           r.price, r.usdprice,
           r.price / u.unit_size                     AS price_per_unit,
           u.base_unit
    FROM raw r JOIN units_df u USING (unit)
    WHERE r.price > 0
""")

LATEST = pd.Timestamp(con.execute("SELECT MAX(month) FROM prices").fetchone()[0])

print(con.execute("""
    SELECT COUNT(*) AS records, COUNT(DISTINCT market) AS markets,
           COUNT(DISTINCT commodity) AS commodities, COUNT(DISTINCT product) AS products,
           COUNT(DISTINCT state) AS states, MIN(date) AS first, MAX(date) AS last
    FROM prices
""").fetchdf().to_string(index=False))

# Why the old "average price" was wrong: the mix of pack sizes changes from year to year
print("\nNaive yearly average price vs share of records that are big wholesale sacks (50–100 kg):")
print(con.execute("""
    SELECT year,
           COUNT(*)                                                         AS records,
           COUNT(DISTINCT state)                                            AS states,
           ROUND(100 * AVG(CASE WHEN unit IN ('100 KG','50 KG') THEN 1 ELSE 0 END)) AS pct_sacks,
           ROUND(AVG(price))                                                AS naive_avg_price
    FROM prices GROUP BY year ORDER BY year
""").fetchdf().to_string(index=False))

# Coverage map: how many markets report in each state and year
cov = con.execute("""
    SELECT state, year, COUNT(DISTINCT market) AS markets
    FROM prices GROUP BY ALL
""").fetchdf().pivot(index="state", columns="year", values="markets")
cov = cov.loc[cov.sum(axis=1).sort_values().index]

fig = px.imshow(cov, text_auto=True, aspect="auto", color_continuous_scale="Blues",
                labels=dict(x="Year", y="State", color="Markets"))
fig.update_layout(
    height=480,
    coloraxis_colorbar=dict(title="<b>Markets<br>surveyed</b>", x=1.02),
    title="<b>1 · Where and when prices were collected</b>"
          "<br><sup>Number of markets reporting per state and year — empty = no data</sup>")
show(fig, "01_coverage",
     "each row is a state, each column a year. The number in a box is how many markets had their prices recorded; "
     "darker blue = more markets.<br>A blank box means no prices at all were collected there that year. "
     "Notice that after 2022 only Borno, Yobe and Adamawa (the North-East) still have data.")

 records  markets  commodities  products  states      first       last
   56163       68           43        67      14 2002-01-15 2026-02-15

Naive yearly average price vs share of records that are big wholesale sacks (50–100 kg):
 year  records  states  pct_sacks  naive_avg_price
 2002       84       2        0.0            189.0
 2003      124       4       10.0            350.0
 2004       28       1      100.0           4300.0
 2005      130       4       21.0           1319.0
 2006       64       4       41.0           1375.0
 2007       51       4       53.0           1672.0
 2008      148       4       18.0           1181.0
 2009      123       5       18.0           1113.0
 2010      193       6       15.0            883.0
 2011      173       5       13.0            866.0
 2012      223       7       32.0           2711.0
 2013      258       9       40.0           3716.0
 2014      977      12       61.0           6694.0
 2015     2029      14       64.0           6872.0
 20

### What we found
- **56,163 prices** from **68 markets** in **14 states**, Jan 2002 – Feb 2026, covering 43 foods sold in
  67 different pack sizes (from a 250 g sachet of salt to a 100 kg sack of maize).
- **Before 2014 the data is very thin** (1–9 states, about 7 foods, mostly wholesale grain). The time series in this
  notebook therefore starts in 2015.
- **Since February 2023 only Borno and Yobe (and Adamawa until late 2025) are surveyed.** Lagos, Kano, Oyo,
  Kaduna and the others each had a single market, and those markets stopped reporting in January 2023.
- The printed table shows why the old "average price" was misleading. It "fell" 80 % from ₦11,338 in 2022 to
  ₦2,289 in 2023. That happened only because 100 kg sacks dropped from 29 % of the records to 3 %.

### What it means for the man in the street
This is the WFP's humanitarian food-price monitoring, and it focuses on the conflict-affected North-East. Recent
numbers describe what a family in Maiduguri or Damaturu pays, not a family in Lagos. It is also **not** the official
NBS inflation figure: the NBS measures a different basket in all 36 states.

## 2. National food price index and inflation

In [2]:
# ================================================================
# 2. NATIONAL RETAIL FOOD PRICE INDEX  (matched-market, chained)
# ================================================================
FOOD_RETAIL = "pricetype = 'Retail' AND category <> 'non-food'"

EVENTS = [("2016-06-01", "Naira float '16"), ("2019-08-01", "Border closure"),
          ("2020-03-01", "COVID-19"), ("2023-06-01", "Petrol subsidy end + naira unified"),
          ("2024-07-01", "Food import duty waiver")]


def chained_index(where=FOOD_RETAIL, group=None, min_products=5):
    """Chained price index built only from like-for-like comparisons.

    1. For every market × product, take the log price change from one observation to the next
       (gaps of up to 3 months are spread evenly over the missing months; jumps > 4.5x are
       treated as data errors; single-month moves are capped at ±50 %).
    2. For every product and month take the mean change across markets. (Markets re-price in
       different months, so the median change is often 0 and would understate inflation.)
    3. The index change is the average across products (every product has equal weight).
    Returns one row per (group,) month with the log change, products used and the index.
    """
    g = f"{group}, " if group else ""
    ch = con.execute(f"""
        WITH m AS (
            SELECT {g}market, product, month, AVG(price) AS p
            FROM prices WHERE {where}
            GROUP BY ALL
        ),
        d AS (
            SELECT *, LN(p / LAG(p) OVER w)                          AS dl,
                   DATE_DIFF('month', LAG(month) OVER w, month)      AS gap
            FROM m
            WINDOW w AS (PARTITION BY market, product ORDER BY month)
        ),
        spread AS (
            SELECT d.*, UNNEST(RANGE(0, gap)) AS k
            FROM d
            WHERE gap BETWEEN 1 AND 3 AND ABS(dl) < 1.5
        )
        SELECT {g}product, (month - TO_MONTHS(k))::DATE AS month,
               GREATEST(LEAST(dl / gap, 0.5), -0.5)     AS dl
        FROM spread
    """).fetchdf()

    keys = [group] if group else []
    prod = ch.groupby(keys + ["product", "month"])["dl"].mean().reset_index()
    agg = (prod.groupby(keys + ["month"])["dl"]
               .agg(change="mean", n_products="count").reset_index())
    agg = agg[agg["n_products"] >= min_products]

    out = []
    for key, grp in (agg.groupby(group) if group else [(None, agg)]):
        grp = grp.set_index("month").sort_index()
        grp = grp.reindex(pd.date_range(grp.index.min(), grp.index.max(), freq="MS"))
        grp["index"] = 100 * np.exp(grp["change"].fillna(0).cumsum())
        if group:
            grp[group] = key
        out.append(grp)
    res = pd.concat(out)
    res.index.name = "month"
    return res


def rebase(s, when="2019-01-01"):
    return 100 * s / s.loc[pd.Timestamp(when)]


nat = chained_index()
nat = nat[nat.index >= "2015-01-01"].copy()          # before 2015 fewer than 5 retail products
nat["index"] = rebase(nat["index"])
nat["yoy"] = (nat["index"] / nat["index"].shift(12) - 1) * 100
nat["mom"] = (np.exp(nat["change"]) - 1) * 100

annual = nat.groupby(nat.index.year).agg(avg_yoy=("yoy", "mean"),
                                         dec_index=("index", "last"),
                                         products=("n_products", "mean")).round(1)
print(annual.to_string())
latest = nat.iloc[-1]
print(f"\nLatest month {nat.index[-1]:%b %Y}: index {latest['index']:.0f} (Jan 2019 = 100), "
      f"YoY {latest['yoy']:+.1f}%")
peak = nat["yoy"].idxmax()
print(f"Peak YoY inflation: {nat.loc[peak, 'yoy']:.1f}% in {peak:%b %Y}")

fig = make_subplots(rows=2, cols=1, shared_xaxes=True, row_heights=[0.55, 0.45],
                    vertical_spacing=0.08,
                    subplot_titles=["Retail food price index (Jan 2019 = 100)",
                                    "Year-on-year food inflation (%)"])
fig.add_trace(go.Scatter(x=nat.index, y=nat["index"], name="Food price level (Jan 2019 = 100)",
                         line=dict(color="#1565C0", width=2.5),
                         hovertemplate="%{x|%b %Y}<br>Index %{y:.0f}<extra></extra>"), row=1, col=1)
fig.add_hline(y=100, line_dash="dash", line_color="grey", row=1, col=1)
legend_item(fig, "Jan 2019 level (100)", "grey", line=True, dash="dash", row=1, col=1)
BANDS = [("Prices falling (below 0 %)", "#2E7D32", -np.inf, 0),
         ("Moderate rise (0–15 %)",     "#F9A825", 0, 15),
         ("High rise (15–30 %)",        "#EF6C00", 15, 30),
         ("Severe rise (above 30 %)",   "#C62828", 30, np.inf)]
for label, colour, lo, hi in BANDS:
    part = nat[(nat["yoy"] > lo) & (nat["yoy"] <= hi)]
    fig.add_trace(go.Bar(x=part.index, y=part["yoy"], marker_color=colour, name=label,
                         legendgroup="yoy", legendgrouptitle_text="Price rise vs 12 months earlier",
                         hovertemplate="%{x|%b %Y}<br>vs a year earlier: %{y:+.1f}%<extra></extra>"), row=2, col=1)
for d, label in EVENTS:
    fig.add_vline(x=pd.Timestamp(d), line_dash="dot", line_color="#6D4C41", line_width=1)
    fig.add_annotation(x=pd.Timestamp(d), y=1.02, yref="paper", text=label, showarrow=False,
                       textangle=-30, font=dict(size=9, color="#6D4C41"), xanchor="left")
legend_item(fig, "Major economic event", "#6D4C41", line=True, dash="dot", row=1, col=1)
fig.update_yaxes(type="log", title_text="Price level", row=1, col=1)
fig.update_yaxes(title_text="% change in a year", ticksuffix="%", row=2, col=1)
fig.update_layout(height=680, barmode="relative", hovermode="x unified", margin=dict(t=120),
                  title="<b>2 · How fast are food prices rising?</b>"
                        "<br><sup>Retail food, matched markets. From Feb 2023 the survey covers only "
                        "Borno, Yobe and Adamawa.</sup>")
show(fig, "02_national_index",
     "Top: the blue line is the cost of a typical basket of food. 100 = its cost in January 2019, so 200 means "
     "twice as expensive, 400 four times.<br>Bottom: each bar is how much dearer food was than 12 months earlier; "
     "the colour shows how serious the rise was (see legend). Green bars below zero mean food got cheaper.<br>"
     "Dotted brown lines mark big events such as the naira float and the end of the petrol subsidy.")

       avg_yoy  dec_index  products
month                              
2015       NaN       70.6      13.2
2016      20.3      103.3      18.9
2017      38.8      120.2      27.2
2018      -7.9      101.2      27.3
2019     -15.0       90.2      34.6
2020      16.7      122.3      35.5
2021      42.7      161.1      35.6
2022      18.1      189.4      23.6
2023      36.8      324.5      22.6
2024      96.8      549.5      25.0
2025       9.5      462.8      25.1
2026     -20.7      438.5      25.5

Latest month Feb 2026: index 438 (Jan 2019 = 100), YoY -20.8%
Peak YoY inflation: 111.6% in Jun 2024


### What we found
- **Food that cost ₦1,000 in January 2019 cost about ₦4,400 in February 2026** (index 438).
- Average food inflation by year: 2016 **20 %**, 2017 **39 %**, 2018 **−8 %**, 2019 **−15 %**, 2020 **17 %**,
  2021 **43 %**, 2022 **18 %**, 2023 **37 %**, **2024 97 %**, 2025 **10 %**.
- The worst moment was **June 2024, when prices were 112 % higher than a year earlier**. In other words, food
  prices more than doubled in twelve months, after the end of the petrol subsidy and the unification of the naira
  in mid-2023.
- **Prices have come down since the peak.** The index fell from 550 (Dec 2024) to 438 (Feb 2026), about 20 % cheaper,
  and in February 2026 prices were 21 % lower than a year earlier.
- The pattern repeats: **big jumps follow naira crises** (2016, 2023–24) and COVID / border closure (2020–21).
  **Falls follow good harvests** (2018–19, and 2025, helped by the July 2024 food import duty waiver).

### What it means for the man in the street
A family that spent **₦20,000 a month on food in 2019 now needs about ₦88,000** for the same food. The recent fall
is real, and it is the first proper relief since 2019. But prices are still more than four times their 2019 level,
so any household whose income has not also risen about four times is eating less, or eating worse, than in 2019.

## 3. Which food groups got more expensive

In [3]:
# ================================================================
# 3. FOOD GROUPS: HOW MUCH PRICES ROSE  (same item, same market)
# ================================================================
L = f"DATE '{LATEST.date()}'"
groups = con.execute(f"""
    WITH m AS (
        SELECT category, market, product,
               MEDIAN(price) FILTER (WHERE year = 2019)                         AS p2019,
               MEDIAN(price) FILTER (WHERE month >  {L} - INTERVAL 12 MONTH)    AS p_now,
               MEDIAN(price) FILTER (WHERE month <= {L} - INTERVAL 12 MONTH
                                       AND month >  {L} - INTERVAL 24 MONTH)    AS p_prev
        FROM prices WHERE {FOOD_RETAIL}
        GROUP BY ALL
    )
    SELECT category,
           100 * (MEDIAN(p_now / p2019) - 1)  AS since_2019_pct,
           100 * (MEDIAN(p_now / p_prev) - 1) AS last_12m_pct,
           COUNT(DISTINCT product)            AS items,
           COUNT(p_now / p2019)               AS market_items
    FROM m GROUP BY category
    ORDER BY since_2019_pct
""").fetchdf()
groups["category"] = groups["category"].str.capitalize()
print("Median price change for the same item in the same market")
print(f"  since_2019 : 2019 average → {LATEST - pd.DateOffset(months=11):%b %Y}–{LATEST:%b %Y}")
print("  last_12m   : previous 12 months → last 12 months\n")
print(groups.round(0).to_string(index=False))

fig = make_subplots(rows=1, cols=2, shared_yaxes=True, horizontal_spacing=0.05,
                    subplot_titles=["Since 2019 (%)", "Last 12 months vs the 12 months before (%)"])
fig.add_trace(go.Bar(y=groups["category"], x=groups["since_2019_pct"], orientation="h", marker_color="#C62828",
                     name="Rise since 2019",
                     text=[f"+{v:.0f}% (x{1 + v / 100:.1f})" for v in groups["since_2019_pct"]],
                     textposition="outside", hovertemplate="%{y}: %{x:+.0f}%<extra></extra>"), row=1, col=1)
for label, colour, mask in [("Rose in the last 12 months", "#EF6C00", groups["last_12m_pct"] > 0),
                            ("Fell in the last 12 months", "#2E7D32", groups["last_12m_pct"] <= 0)]:
    part = groups[mask]
    fig.add_trace(go.Bar(y=part["category"], x=part["last_12m_pct"], orientation="h", marker_color=colour,
                         name=label, text=part["last_12m_pct"].map("{:+.0f}%".format), textposition="outside",
                         hovertemplate="%{y}: %{x:+.0f}%<extra></extra>"), row=1, col=2)
fig.update_xaxes(ticksuffix="%")
fig.update_yaxes(categoryorder="array", categoryarray=groups["category"].tolist())
fig.update_layout(height=440, barmode="relative", margin=dict(l=170),
                  title="<b>3 · Which food groups got more expensive</b>"
                        "<br><sup>Retail, same item in the same market (recent period = North-East markets)</sup>")
show(fig, "03_food_groups",
     "Left: how much more each food group costs now than in 2019. +400% (x5.0) means five times the 2019 price.<br>"
     "Right: the change over the last year only. Orange = still getting dearer, green = getting cheaper.",
     legend_title="Colour key")

Median price change for the same item in the same market
  since_2019 : 2019 average → Mar 2025–Feb 2026
  last_12m   : previous 12 months → last 12 months

             category  since_2019_pct  last_12m_pct  items  market_items
Vegetables and fruits           286.0           0.0      6            93
       Milk and dairy           300.0           1.0      1            15
   Cereals and tubers           376.0         -25.0     14            92
   Miscellaneous food           395.0           1.0      3            31
      Pulses and nuts           419.0         -25.0      7            63
         Oil and fats           447.0          17.0      2            31
  Meat, fish and eggs           449.0          10.0      4            64


### What we found
- Since 2019, prices have risen about **4 to 5.5 times in every food group**: meat, fish & eggs **+449 %**,
  cooking oil **+447 %**, beans/cowpeas/groundnuts **+419 %**, sugar/salt/seasoning **+395 %**,
  grains & tubers **+376 %**, milk **+300 %**, vegetables & fruit **+286 %**.
- **Last 12 months:** grains & tubers **−25 %**, beans & groundnuts **−25 %**. Vegetables, milk and
  sugar/salt were flat. **Cooking oil +17 %** and **meat, fish & eggs +10 %** are still rising.

### What it means for the man in the street
The 2025 relief is mostly in **grains and beans**, the cheap foods that fill the stomach. **Protein (meat, fish,
eggs) and cooking oil keep getting dearer.** Families can afford to eat enough, but less nutritious food. That is
a particular risk for children and pregnant women.

## 4. Then and now — the price of everyday items

In [4]:
# ================================================================
# 4. ITEM PRICES: 2019 vs LAST 12 MONTHS (same markets only)
# ================================================================
items = con.execute(f"""
    WITH m AS (
        SELECT product, commodity, unit, base_unit, market,
               MEDIAN(CASE WHEN year = 2019 THEN price END)                   AS p2019,
               MEDIAN(CASE WHEN month > DATE '{LATEST.date()}' - INTERVAL 12 MONTH
                           THEN price END)                                     AS p_now
        FROM prices
        WHERE {FOOD_RETAIL}
        GROUP BY ALL
    )
    SELECT commodity, unit, product,
           COUNT(*)            AS markets,
           MEDIAN(p2019)       AS price_2019,
           MEDIAN(p_now)       AS price_now,
           MEDIAN(p_now / p2019) AS ratio
    FROM m
    WHERE p2019 IS NOT NULL AND p_now IS NOT NULL
    GROUP BY ALL
    HAVING COUNT(*) >= 3
    ORDER BY ratio DESC
""").fetchdf()
items["pct_change"] = (items["ratio"] - 1) * 100
print(f"Comparison window: 2019 vs {LATEST - pd.DateOffset(months=11):%b %Y}–{LATEST:%b %Y}\n")
print(items[["commodity", "unit", "markets", "price_2019", "price_now", "pct_change"]]
      .round(0).to_string(index=False))
print(f"\nMedian item: prices x{items['ratio'].median():.1f}")

items = items.sort_values("pct_change")
fig = go.Figure()
for label, colour, lo, hi in [("Less than doubled (under +100 %)", "#FDD49E", -np.inf, 100),
                              ("2 to 4 times dearer (+100 to +300 %)", "#FC8D59", 100, 300),
                              ("4 to 5 times dearer (+300 to +400 %)", "#D7301F", 300, 400),
                              ("More than 5 times dearer (over +400 %)", "#7F0000", 400, np.inf)]:
    part = items[(items["pct_change"] > lo) & (items["pct_change"] <= hi)]
    fig.add_trace(go.Bar(
        x=part["pct_change"], y=part["commodity"] + " (" + part["unit"] + ")", orientation="h",
        marker_color=colour, name=label,
        text=[f"₦{a:,.0f} → ₦{b:,.0f}" for a, b in zip(part["price_2019"], part["price_now"])],
        textposition="outside",
        hovertemplate="<b>%{y}</b><br>Change %{x:+.0f}%<br>%{text}<extra></extra>"))
fig.update_yaxes(categoryorder="array", categoryarray=(items["commodity"] + " (" + items["unit"] + ")").tolist())
fig.update_layout(height=max(500, 26 * len(items) + 140), margin=dict(l=220), barmode="relative",
                  xaxis=dict(title="Price change since 2019 (%)", ticksuffix="%"),
                  title="<b>4 · What the same item cost in 2019 vs today</b>"
                        "<br><sup>Retail, median across the same markets (mostly North-East). "
                        "Label = typical price per pack, 2019 → last 12 months.</sup>")
show(fig, "04_item_prices",
     "each bar is one everyday item in the pack size shown in brackets. The longer the bar, the more its price has "
     "risen since 2019.<br>The label at the end of the bar gives the actual shop price: 2019 price → price over the "
     "last 12 months. The colour groups items by how many times dearer they are.",
     legend_title="How much dearer since 2019")

Comparison window: 2019 vs Mar 2025–Feb 2026

      commodity   unit  markets  price_2019  price_now  pct_change
            Yam 2.5 KG       16       403.0     3000.0       644.0
    Maize flour 2.1 KG       16       242.0     1807.0       537.0
Oil (vegetable)      L       15       427.0     2750.0       533.0
     Groundnuts 2.2 KG       16       596.0     3596.0       497.0
           Eggs 30 pcs       16      1025.0     6000.0       479.0
          Sugar  500 G       16       173.0     1000.0       476.0
           Fish     KG       16       193.0     1393.0       462.0
    Watermelons 2.1 KG       15       200.0     1034.0       441.0
        Bananas 1.3 KG       16       177.0      757.0       420.0
        Cowpeas 2.5 KG       15       481.0     2500.0       419.0
     Oil (palm)      L       16       480.0     2481.0       403.0
    Beans (red) 2.5 KG       16       521.0     2500.0       397.0
    Meat (beef)     KG       16      1150.0     5516.0       397.0
  Beans (white) 

### What we found
Typical prices in the same markets, 2019 → last 12 months (the change is the median across markets):

| Item | 2019 | Now | Change |
|---|---|---|---|
| Crate of 30 eggs | ₦1,025 | ₦6,000 | ×5.8 |
| 1 litre vegetable oil | ₦427 | ₦2,750 | ×6.3 |
| 2.1 kg maize flour | ₦242 | ₦1,807 | ×6.4 |
| 2.5 kg yam | ₦403 | ₦3,000 | ×7.4 |
| 1 kg beef | ₦1,150 | ₦5,516 | ×5.0 |
| 2.8 kg imported rice | ₦1,000 | ₦4,741 | ×4.8 |
| 2.7 kg local rice | ₦680 | ₦2,770 | ×4.1 |
| 2.6 kg millet | ₦232 | ₦1,147 | ×4.5 |
| 0.5 kg tomatoes | ₦72 | ₦200 | ×2.7 |
| 250 g salt | ₦175 | ₦250 | ×1.7 |

The typical (median) item costs **5 times** its 2019 price.

### What it means for the man in the street
- **Processed and factory foods rose the most:** maize flour, vegetable oil, sugar, and eggs (poultry feed is
  imported and priced in dollars). **Local fresh produce rose the least:** spinach, tomatoes, onions.
- **Maize flour rose ×6.4, but whole millet and sorghum grain rose only about ×4.5.** Buying grain and milling it
  locally is much cheaper than buying packaged flour.
- Local rice (×4.1) held its price better than imported rice (×4.8).

## 5. Where is food most expensive, and where did it rise fastest?

In [5]:
# ================================================================
# 5. STATE PRICE LEVELS & INFLATION  (period when all states report)
# ================================================================
P0, P1 = "2021-01-01", "2023-01-01"      # 13 states report continuously in this window

levels = con.execute(f"""
    WITH sm AS (   -- state median price per product-month
        SELECT state, product, month, MEDIAN(price) AS p
        FROM prices
        WHERE {FOOD_RETAIL} AND month BETWEEN DATE '{P0}' AND DATE '{P1}'
        GROUP BY ALL
    ),
    rel AS (       -- compared with the median of all states for the SAME product and month
        SELECT state, LN(p / MEDIAN(p) OVER (PARTITION BY product, month)) AS lr
        FROM sm
    )
    SELECT state, 100 * (EXP(AVG(lr)) - 1) AS price_level_pct, COUNT(*) AS comparisons
    FROM rel GROUP BY state HAVING COUNT(*) >= 100
""").fetchdf().set_index("state")

st_idx = chained_index(group="state", min_products=3)
st_idx = st_idx[(st_idx.index >= P0) & (st_idx.index <= P1)]
infl = st_idx.groupby("state")["change"].agg(lambda s: 100 * (np.exp(s.sum()) - 1) if s.notna().sum() >= 20 else np.nan)
levels["inflation_pct"] = infl
levels = levels.dropna().sort_values("price_level_pct")
print(f"Window {P0[:7]} → {P1[:7]}")
print(levels.round(1).to_string())

fig = make_subplots(rows=1, cols=2, shared_yaxes=True, horizontal_spacing=0.04,
                    subplot_titles=["Price level vs all-state median (%)",
                                    f"Food price rise {P0[:4]}–{P1[:4]} (%)"])
for label, colour, mask in [("Dearer than the typical state", "#C62828", levels["price_level_pct"] > 0),
                            ("Cheaper than the typical state", "#2E7D32", levels["price_level_pct"] <= 0)]:
    part = levels[mask]
    fig.add_trace(go.Bar(y=part.index, x=part["price_level_pct"], orientation="h", marker_color=colour, name=label,
                         text=part["price_level_pct"].map("{:+.0f}%".format), textposition="outside",
                         hovertemplate="%{y}: %{x:+.1f}%<extra></extra>"), row=1, col=1)
fig.add_trace(go.Bar(y=levels.index, x=levels["inflation_pct"], orientation="h", marker_color="#EF6C00",
                     name=f"Price rise {P0[:4]}–{P1[:4]}",
                     text=levels["inflation_pct"].map("{:.0f}%".format), textposition="outside",
                     hovertemplate="%{y}: %{x:.1f}%<extra></extra>"), row=1, col=2)
fig.update_yaxes(categoryorder="array", categoryarray=levels.index.tolist())
fig.update_xaxes(ticksuffix="%")
fig.update_layout(height=520, barmode="relative",
                  title="<b>5 · Food prices by state</b>"
                        f"<br><sup>Same products compared in the same month, {P0[:7]} to {P1[:7]} "
                        "(the last period with all states in the survey)</sup>")
show(fig, "05_states",
     "Left: how much more (red) or less (green) the same food costs in each state compared with the typical state. "
     "Lagos +30% means<br>an item costing ₦1,000 in the typical state costs about ₦1,300 in Lagos. "
     "Right: how much food prices went up in each state over those two years.",
     legend_title="Colour key")

Window 2021-01 → 2023-01
         price_level_pct  comparisons  inflation_pct
state                                               
Jigawa             -12.1          198           64.6
Yobe                -9.4          648           48.1
Kano                -6.8          234           46.4
Zamfara             -6.0          221           36.0
Katsina             -5.1          221           38.1
Borno               -0.8          664           45.8
Kaduna              -0.4          309           15.1
Kebbi                1.3          236           57.9
Adamawa              6.2          235           48.5
Oyo                 11.0          244           60.4
Gombe               14.3          253           36.2
Abia                22.0          204           30.6
Lagos               29.8          235           25.2


### What we found
Jan 2021 – Jan 2023 was the last period in which all 13 states were surveyed:
- **Most expensive:** Lagos **+30 %** above the typical state for the *same item in the same month*,
  Abia **+22 %**, Gombe **+14 %**, Oyo **+11 %**.
- **Cheapest:** Jigawa **−12 %**, Yobe **−9 %**, Kano **−7 %**, Zamfara **−6 %**, Katsina **−5 %**.
- **But the cheap northern states saw the fastest increases:** Jigawa **+65 %** in two years, Kebbi +58 %,
  Adamawa +49 %, Yobe +48 %, while Lagos rose +25 % and Kaduna +15 %.

### What it means for the man in the street
Food is cheapest where it is grown (the North) and dearest in the southern cities, where transport, rent and
extra middlemen are added: Lagos pays about 30 % more for the same item. That gap is closing fast. Rural northern
households, who are usually the poorest and spend most of their income on food, are taking the biggest price
increases. (Most states have only one surveyed market, so treat the state figures as indicative.)

## 6. The seasonal calendar — when to buy, when to sell

In [6]:
# ================================================================
# 6. SEASONAL PRICE PATTERN BY FOOD  (trend removed)
# ================================================================
FAMILIES = {"Maize": "maize", "Millet": "millet", "Sorghum": "sorghum", "Rice (local)": r"rice \(local|rice \(milled",
            "Beans & cowpeas": "beans|cowpea", "Tomatoes": "tomato", "Onions": "onion", "Yam": "yam"}
MONTHS = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]

raw_s = con.execute("""
    SELECT market, product, commodity, month, AVG(price) AS p
    FROM prices WHERE category <> 'non-food' AND commodity NOT ILIKE '%flour%'
    GROUP BY ALL
""").fetchdf()
raw_s["family"] = None
for fam, pat in FAMILIES.items():
    raw_s.loc[raw_s["family"].isna() & raw_s["commodity"].str.lower().str.contains(pat), "family"] = fam
raw_s = raw_s.dropna(subset=["family"])

rows = []
for (market, product, family), s in raw_s.groupby(["market", "product", "family"]):
    s = s.set_index("month")["p"]
    s.index = pd.DatetimeIndex(s.index)
    if len(s) < 30:
        continue
    lp = np.log(s).reindex(pd.date_range(s.index.min(), s.index.max(), freq="MS"))
    trend = lp.interpolate(limit=2).rolling(12).mean().rolling(2).mean().shift(-6)   # centred 2x12 MA
    dev = (lp - trend).dropna()
    rows.append(pd.DataFrame({"family": family, "cal_month": dev.index.month, "dev": dev.values}))
seas = pd.concat(rows)
prof = seas.groupby(["family", "cal_month"])["dev"].median().unstack()
prof = prof.sub(prof.mean(axis=1), axis=0)
prof = (np.exp(prof) - 1) * 100                              # % above / below the yearly average
prof.columns = MONTHS

summary = pd.DataFrame({"cheapest": prof.idxmin(axis=1), "dearest": prof.idxmax(axis=1),
                        "swing_pct": prof.max(axis=1) - prof.min(axis=1),
                        "series": seas.groupby("family").size()}).round(0)
print(summary.sort_values("swing_pct", ascending=False).to_string())

fig = go.Figure()
for i, fam in enumerate(FAMILIES):
    if fam in prof.index:
        fig.add_trace(go.Scatter(x=MONTHS, y=prof.loc[fam], name=fam, mode="lines+markers",
                                 line=dict(width=2.2, color=px.colors.qualitative.Bold[i]),
                                 visible=True if fam in ("Maize", "Millet", "Sorghum", "Beans & cowpeas", "Tomatoes") else "legendonly"))
fig.add_hline(y=0, line_dash="dash", line_color="grey")
fig.add_vrect(x0="Jun", x1="Aug", fillcolor="#C62828", opacity=0.07, line_width=0,
              annotation_text="Lean season", annotation_position="top left")
fig.add_vrect(x0="Oct", x1="Dec", fillcolor="#2E7D32", opacity=0.07, line_width=0,
              annotation_text="Harvest", annotation_position="top left")
legend_item(fig, "Yearly average price (0 %)", "grey", line=True, dash="dash")
legend_item(fig, "Lean season: dearest (Jun–Aug)", "rgba(198,40,40,0.25)")
legend_item(fig, "Harvest: cheapest (Oct–Dec)", "rgba(46,125,50,0.25)")
fig.update_layout(height=470, hovermode="x unified", yaxis=dict(title="% above / below yearly average", ticksuffix="%"),
                  title="<b>6 · Seasonal price calendar</b>"
                        "<br><sup>Typical price in each month relative to the yearly average, long-term trend removed. "
                        "Click legend items to add Rice, Onions, Yam.</sup>")
show(fig, "06_seasonality",
     "each line is one food. A point above the grey dashed line means that food is usually dearer than normal in "
     "that month; below means cheaper.<br>Example: a food at +12% in August and −10% in November costs about 25% "
     "more in August than in November. Click a food in the legend to show or hide it.",
     legend_title="Food")

                cheapest dearest  swing_pct  series
family                                             
Tomatoes             Mar     Jul       31.0     179
Maize                Nov     Aug       25.0    1589
Sorghum              Dec     Aug       24.0    2064
Onions               Feb     Dec       23.0     458
Yam                  Feb     Jul       20.0    1035
Millet               Oct     Aug       20.0    1351
Beans & cowpeas      Dec     Aug       17.0    2300
Rice (local)         Dec     Aug        8.0    1451


### What we found
Prices follow the farming calendar every year (long-term inflation removed):

| Food | Cheapest | Dearest | Swing |
|---|---|---|---|
| Maize | November | August | 25 % |
| Sorghum | December | August | 24 % |
| Millet | October | August | 20 % |
| Beans & cowpeas | December | August | 17 % |
| Local rice | December | August | 8 % |
| Tomatoes | March | July | 31 % |
| Yam | February | July | 20 % |
| Onions | February | December | 23 % |

### What it means for the man in the street
- **Households:** buy your grain and beans **in bulk between October and December**, after the harvest. The same bag
  costs about **20–25 % more in July–August** (the "lean season", when last year's harvest is running out).
- **Farmers and traders:** if you can store grain safely (hermetic / PICS bags), selling in June–August instead of
  at harvest earns about 20 % more.
- **Tomatoes are cheapest in February–March.** That is the time to dry them or make paste.
- **Government:** June–August is the right time to release strategic grain reserves.

## 7. Retail vs wholesale — how much the last step to the consumer adds

In [7]:
# ================================================================
# 7. RETAIL MARK-UP OVER WHOLESALE  (same market, month, commodity, per kg)
# ================================================================
markup = con.execute("""
    WITH r AS (SELECT market, month, commodity, MEDIAN(price_per_unit) AS rp
               FROM prices WHERE pricetype = 'Retail' AND base_unit = 'kg' GROUP BY ALL),
         w AS (SELECT market, month, commodity, MEDIAN(price_per_unit) AS wp
               FROM prices WHERE pricetype = 'Wholesale' AND base_unit = 'kg' GROUP BY ALL)
    SELECT commodity, COUNT(*) AS pairs,
           100 * (MEDIAN(rp / wp) - 1)              AS markup_pct,
           100 * (QUANTILE_CONT(rp / wp, 0.25) - 1) AS q25,
           100 * (QUANTILE_CONT(rp / wp, 0.75) - 1) AS q75,
           MEDIAN(wp) AS wholesale_per_kg, MEDIAN(rp) AS retail_per_kg
    FROM r JOIN w USING (market, month, commodity)
    WHERE rp / wp BETWEEN 0.5 AND 4                  -- drop obvious unit / entry errors
    GROUP BY commodity HAVING COUNT(*) >= 30
    ORDER BY markup_pct
""").fetchdf()
print(markup.round(0).to_string(index=False))
print(f"\nMedian mark-up across commodities: {markup['markup_pct'].median():.0f}%")

fig = go.Figure()
for label, colour, mask in [("Small extra cost (under 20 %)", "#7E57C2", markup["markup_pct"] < 20),
                            ("Large extra cost (20 % or more)", "#4A148C", markup["markup_pct"] >= 20)]:
    part = markup[mask]
    fig.add_trace(go.Bar(
        x=part["markup_pct"], y=part["commodity"], orientation="h", marker_color=colour, name=label,
        error_x=dict(type="data", array=part["q75"] - part["markup_pct"],
                     arrayminus=part["markup_pct"] - part["q25"], color="rgba(0,0,0,0.35)"),
        text=[f"₦{w:,.0f} → ₦{r:,.0f}/kg" for w, r in zip(part["wholesale_per_kg"], part["retail_per_kg"])],
        textposition="outside",
        hovertemplate="<b>%{y}</b><br>Mark-up %{x:.0f}%<br>%{text}<extra></extra>"))
legend_item(fig, "Thin black line = usual range", "rgba(0,0,0,0.5)", line=True)
fig.update_yaxes(categoryorder="array", categoryarray=markup["commodity"].tolist())
fig.update_layout(height=460, margin=dict(l=200), barmode="relative",
                  xaxis=dict(title="Retail price above wholesale (%)", ticksuffix="%"),
                  title="<b>7 · Retail mark-up over wholesale, per kg</b>"
                        "<br><sup>Same market and month. Bar = median, whisker = middle 50 % of months. "
                        "Label = median wholesale → retail price per kg.</sup>")
show(fig, "07_markup",
     "each bar shows how much more you pay per kg when you buy in small measures at retail instead of by the "
     "bag (wholesale) in the same market.<br>0% = no difference. The label gives the typical price per kg: "
     "wholesale → retail. The thin black line shows the range seen in most months.",
     legend_title="Extra cost of buying small")

                  commodity  pairs  markup_pct   q25  q75  wholesale_per_kg  retail_per_kg
               Rice (local)    351        -3.0 -10.0  6.0             380.0          383.0
                     Millet    356         2.0  -5.0 10.0             198.0          198.0
              Maize (white)    253         2.0  -4.0 12.0             187.0          175.0
             Maize (yellow)    285         3.0  -4.0 12.0             186.0          188.0
            Rice (imported)    262         4.0  -5.0 12.0             524.0          509.0
            Sorghum (white)    259         4.0  -5.0 15.0             178.0          184.0
            Sorghum (brown)    322         5.0  -7.0 18.0             182.0          185.0
            Cowpeas (white)    264         7.0  -1.0 17.0             340.0          325.0
            Cowpeas (brown)    262         9.0  -2.0 25.0             364.0          359.0
       Groundnuts (shelled)    283        10.0  -6.0 22.0             400.0          435.0



Median mark-up across commodities: 5%


### What we found
In the same market and month, per kg:
- For **grains, rice, beans and groundnuts**, the retail price is almost the same as the wholesale price (median
  mark-up **−3 % to +10 %**).
- **Gari is the exception: +47 % (yellow) to +54 % (white)** more when bought retail in small measures.

### What it means for the man in the street
- In these markets, buying grain in small measures does **not** cost much more per kg. The poor, who can only buy
  a little at a time, are not heavily penalised for grain.
- **Gari bought by the cup costs about 50 % more than gari bought by the bag.** Buying a bag, or sharing one with
  neighbours, cuts the gari bill by up to a third.
- Traders' margins are thin. **The huge price rises did not come from retailers.** They came from further up the
  chain: the cost of producing, transporting and importing food, and the value of the naira.
- (Wholesale prices were only collected until 2023.)

## 8. What pushes food prices — the naira and petrol

In [8]:
# ================================================================
# 8. DRIVERS: EXCHANGE RATE & PETROL PRICE vs FOOD INDEX
# ================================================================
drivers = con.execute("""
    SELECT month,
           MEDIAN(price / usdprice) FILTER (WHERE usdprice > 0)                          AS ngn_per_usd,
           MEDIAN(price_per_unit)   FILTER (WHERE commodity = 'Fuel (petrol-gasoline)'
                                            AND pricetype = 'Retail')                    AS petrol_per_litre
    FROM prices GROUP BY month ORDER BY month
""").fetchdf().set_index("month")
drivers.index = pd.DatetimeIndex(drivers.index)
drivers = drivers.reindex(nat.index).interpolate(limit=3)
drivers["food_index"] = nat["index"]

reb = drivers.apply(lambda s: rebase(s))
reb["food_in_usd"] = reb["food_index"] / reb["ngn_per_usd"] * 100

yoy = np.log(drivers).diff(12)
lag_corr = pd.DataFrame({
    driver: [yoy["food_index"].corr(yoy[driver].shift(k)) for k in range(0, 13)]
    for driver in ["ngn_per_usd", "petrol_per_litre"]}, index=range(0, 13)).round(2)
lag_corr.index.name = "lag (months)"
print("Correlation of food YoY change with driver YoY change, driver leading by k months:")
print(lag_corr.T.to_string())
for d in lag_corr:
    print(f"  strongest link for {d}: {lag_corr[d].max():.2f} at {lag_corr[d].idxmax()} months")

print(f"\nJan 2019 → {nat.index[-1]:%b %Y}:  naira per $ x{drivers['ngn_per_usd'].iloc[-1] / drivers.loc['2019-01-01', 'ngn_per_usd']:.1f}, "
      f"petrol x{drivers['petrol_per_litre'].iloc[-1] / drivers.loc['2019-01-01', 'petrol_per_litre']:.1f}, "
      f"food x{drivers['food_index'].iloc[-1] / 100:.1f}")

fig = go.Figure()
for col, name, colour, dash in [("food_index", "Food prices in naira", "#1565C0", None),
                                ("ngn_per_usd", "Naira needed to buy 1 US dollar", "#C62828", None),
                                ("petrol_per_litre", "Petrol price per litre", "#EF6C00", None),
                                ("food_in_usd", "Food prices in US dollars", "#1565C0", "dot")]:
    fig.add_trace(go.Scatter(x=reb.index, y=reb[col], name=name, line=dict(color=colour, width=2.3, dash=dash),
                             hovertemplate=f"{name}<br>%{{x|%b %Y}}: %{{y:.0f}}<extra></extra>"))
fig.add_hline(y=100, line_dash="dash", line_color="grey")
legend_item(fig, "Jan 2019 level (100)", "grey", line=True, dash="dash")
fig.update_yaxes(type="log", title="Jan 2019 = 100 (log scale)")
fig.update_layout(height=500, hovermode="x unified",
                  title="<b>8 · Food prices follow the naira and petrol</b>"
                        "<br><sup>All series rebased to Jan 2019 = 100. Exchange rate implied by WFP's USD prices. "
                        "Petrol was not recorded Feb–Dec 2025.</sup>")
show(fig, "08_drivers",
     "every line starts at 100 in January 2019, so 400 means four times the January 2019 level. The solid blue line "
     "(food) and the red line<br>(the dollar exchange rate) climb together: when the naira weakens, food follows. "
     "The dotted blue line is food priced in dollars; it stays near 100,<br>meaning food has not really become dearer "
     "in dollar terms — the naira has lost value.",
     legend_title="Line")

Correlation of food YoY change with driver YoY change, driver leading by k months:
lag (months)        0     1     2     3     4     5     6     7     8     9     10    11    12
ngn_per_usd       0.79  0.81  0.81  0.80  0.77  0.73  0.67  0.60  0.51  0.41  0.31  0.23  0.13
petrol_per_litre  0.49  0.47  0.49  0.51  0.54  0.53  0.51  0.48  0.47  0.47  0.46  0.45  0.43
  strongest link for ngn_per_usd: 0.81 at 1 months
  strongest link for petrol_per_litre: 0.54 at 4 months

Jan 2019 → Feb 2026:  naira per $ x4.4, petrol x6.5, food x4.4


### What we found
- From Jan 2019 to Feb 2026: **naira per US dollar ×4.4, petrol ×6.5, food ×4.4**.
- Food inflation moves closely with the fall of the naira: correlation **0.81, with food following about 1 month
  later**. The link with petrol is weaker (0.54, about 4 months later).
- **Measured in US dollars, food costs about the same as in 2019** (the dotted line stays near 100).

### What it means for the man in the street
Nigeria's food inflation is mostly **the naira losing value**, not food becoming scarcer. Fertiliser, fuel,
machinery, wheat, feed and packaging are all priced in dollars. **When the naira falls, expect food prices to follow
within one to three months.** Stocking up on dry goods (rice, beans, oil) when the naira starts to slide protects
the family budget. For government, a stable exchange rate is probably the strongest food-price policy available.
(Correlation is not proof of cause, but the timing and size of the link are consistent over seven years.)

## 9. Forecast — where are food prices heading in the next 6 months?

In [9]:
# ================================================================
# 9. FORECASTING THE FOOD INDEX — honest back-test vs simple benchmarks
# ================================================================
from sklearn.linear_model import Ridge
from sklearn.neural_network import MLPRegressor

H = 6                                                      # forecast horizon (months)
y = np.log(nat["index"])                                   # model the log index
dy = y.diff()


def features(d, t):
    """Features available at the END of month t, to predict the change in month t+1."""
    nxt = (t + pd.DateOffset(months=1)).month
    return [d.loc[:t].iloc[-1], d.loc[:t].iloc[-3:].mean(), d.loc[:t].iloc[-12:].mean(),
            np.sin(2 * np.pi * nxt / 12), np.cos(2 * np.pi * nxt / 12)]


def fit_predict(kind, d_hist, h=H):
    """Recursive h-step forecast of monthly log changes, using only d_hist."""
    if kind == "Naive (no change)":
        return np.zeros(h)
    if kind == "Trend (last 12m pace)":
        return np.repeat(d_hist.iloc[-12:].mean(), h)
    months = d_hist.index[13:]
    X = np.array([features(d_hist, m - pd.DateOffset(months=1)) for m in months])
    t = d_hist.loc[months].values
    model = (Ridge(alpha=1.0) if kind == "Ridge AR" else
             MLPRegressor(hidden_layer_sizes=(16,), alpha=1e-2, max_iter=3000, random_state=0))
    model.fit(X, t)
    d = d_hist.copy()
    out = []
    for _ in range(h):
        t_end = d.index[-1]
        step = model.predict([features(d, t_end)])[0]
        out.append(step)
        d.loc[t_end + pd.DateOffset(months=1)] = step
    return np.array(out)


MODELS = ["Naive (no change)", "Trend (last 12m pace)", "Ridge AR", "MLP neural net"]
d_all = dy.dropna()
origins = [m for m in d_all.index if pd.Timestamp("2019-01-01") <= m <= d_all.index[-1] - pd.DateOffset(months=H)]
errs = []
for o in origins:
    hist = d_all.loc[:o]
    actual = y.loc[o + pd.DateOffset(months=1): o + pd.DateOffset(months=H)].values - y.loc[o]
    for kind in MODELS:
        pred = np.cumsum(fit_predict(kind, hist))
        for h, (a, p) in enumerate(zip(actual, pred), 1):
            errs.append(dict(model=kind, origin=o, h=h, err=p - a))
errs = pd.DataFrame(errs)
errs["abs_pct"] = 100 * (np.exp(errs["err"].abs()) - 1)

score = errs.pivot_table(index="model", columns="h", values="abs_pct", aggfunc="mean").round(1)
score.columns = [f"{h}m ahead" for h in score.columns]
score = score.loc[score.mean(axis=1).sort_values().index]
print(f"Back-test: {len(origins)} forecast origins ({origins[0]:%b %Y} – {origins[-1]:%b %Y})")
print("Mean absolute error of the forecast index (%) — lower is better:")
print(score.to_string())

best = score.index[0]
print(f"\nBest model on back-test: {best}")

# Forecast from the latest month with the best model; interval from its own back-test errors
path = np.cumsum(fit_predict(best, d_all))
q = errs[errs["model"] == best].groupby("h")["err"].quantile([0.1, 0.9]).unstack()
fdates = pd.date_range(nat.index[-1] + pd.DateOffset(months=1), periods=H, freq="MS")
base = nat["index"].iloc[-1]
fc = pd.DataFrame({"forecast": base * np.exp(path),
                   "low_80": base * np.exp(path - q[0.9].values),
                   "high_80": base * np.exp(path - q[0.1].values)}, index=fdates)
print(fc.round(0).to_string())
print(f"\n{H}-month outlook: {100 * (fc['forecast'].iloc[-1] / base - 1):+.1f}% "
      f"(80% range {100 * (fc['low_80'].iloc[-1] / base - 1):+.0f}% to {100 * (fc['high_80'].iloc[-1] / base - 1):+.0f}%)")

fig = make_subplots(rows=1, cols=2, column_widths=[0.62, 0.38], horizontal_spacing=0.1,
                    subplot_titles=[f"Food index and {H}-month forecast ({best})",
                                    "Back-test error by horizon"])
hist_plot = nat.loc["2021-01-01":, "index"]
fig.add_trace(go.Scatter(x=hist_plot.index, y=hist_plot, name="Actual food price level",
                         legendgroup="fc", legendgrouptitle_text="Left: the forecast",
                         line=dict(color="#1565C0", width=2.5)), row=1, col=1)
fig.add_trace(go.Scatter(x=list(fdates) + list(fdates[::-1]), y=list(fc["high_80"]) + list(fc["low_80"][::-1]),
                         fill="toself", fillcolor="rgba(198,40,40,0.12)", line=dict(width=0), legendgroup="fc",
                         name="Likely range (8 in 10 chance)", hoverinfo="skip"), row=1, col=1)
fig.add_trace(go.Scatter(x=[hist_plot.index[-1]] + list(fdates), y=[base] + list(fc["forecast"]), legendgroup="fc",
                         name="Most likely path", line=dict(color="#C62828", width=2.5, dash="dash")), row=1, col=1)
for i, m in enumerate(score.index):
    fig.add_trace(go.Scatter(x=list(range(1, H + 1)), y=score.loc[m].values, name=m, mode="lines+markers",
                             legendgroup="err", legendgrouptitle_text="Right: forecast method",
                             line=dict(color=px.colors.qualitative.Bold[i], width=2)), row=1, col=2)
fig.update_xaxes(title_text="Months ahead", row=1, col=2)
fig.update_yaxes(title_text="Index (Jan 2019 = 100)", row=1, col=1)
fig.update_yaxes(title_text="Average miss (%)", ticksuffix="%", row=1, col=2)
fig.update_layout(height=480, legend=dict(groupclick="toggleitem"),
                  title="<b>9 · Food price outlook</b>"
                        "<br><sup>Each model was re-trained at every month since 2019 using only past data, "
                        "then scored on what really happened.</sup>")
show(fig, "09_forecast",
     "Left: the blue line is what food prices actually did; the red dashed line is the expected path for the next "
     "6 months, and the shaded band<br>is where prices will land 8 times out of 10. Right: how far off each forecasting "
     "method was on average when tested on past years —<br>lower is better. Errors grow the further ahead you look.")

Back-test: 80 forecast origins (Jan 2019 – Aug 2025)
Mean absolute error of the forecast index (%) — lower is better:
                       1m ahead  2m ahead  3m ahead  4m ahead  5m ahead  6m ahead
model                                                                            
Ridge AR                    2.7       5.1       7.3       9.6      12.1      14.6
Trend (last 12m pace)       2.8       5.1       7.4       9.9      12.8      15.8
Naive (no change)           3.3       6.1       8.9      11.7      14.5      17.6
MLP neural net              5.8      10.5      14.9      18.3      21.4      24.4

Best model on back-test: Ridge AR
            forecast  low_80  high_80
2026-03-01     448.0   433.0    471.0
2026-04-01     461.0   432.0    501.0
2026-05-01     475.0   427.0    536.0
2026-06-01     487.0   424.0    571.0
2026-07-01     496.0   420.0    605.0
2026-08-01     501.0   414.0    633.0

6-month outlook: +14.2% (80% range -5% to +44%)


### What we found
- Each model was re-trained **80 times** (each month from Jan 2019 to Aug 2025) using only the data available at the
  time. It was then scored on what actually happened next.
- Average error: **Ridge AR 2.7 %** one month ahead, **7.3 %** three months ahead, **14.6 %** six months ahead.
  The simple "continue the last 12 months' pace" rule is almost as good. "No change" is worse (17.6 % at 6 months).
- **The neural network (MLP) was the worst model** (24 % error at 6 months). There are only about 130 months of data,
  which is far too little for a neural network. The previous notebook's "LSTM" claims were not supported.
- **Outlook: the index rises from 438 to about 501 by August 2026, i.e. +14 %** (80 % range: −5 % to +44 %). Most of
  this is the normal lean-season rise.

### What it means for the man in the street
Expect food prices to **climb again from now until about August 2026**, roughly 14 %, and then ease after the
harvest. The range is wide: a new naira shock could make it +40 %, while a good season could keep prices flat.
**Buy the family's staples now rather than in June–August.** Nobody, including this model, can predict food prices
precisely more than a few months ahead.

## 10. Early warning — can we see a price surge coming?

In [10]:
# ================================================================
# 10. EARLY-WARNING MODEL: P(state food prices rise > 10% in the next 3 months)
# ================================================================
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import precision_score, recall_score, roc_auc_score, roc_curve
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler

SURGE = np.log(1.10)            # a "surge" = +10 % or more over the next 3 months
SPLIT = pd.Timestamp("2021-01-01")

st = chained_index(group="state", min_products=3).reset_index()
st = st.sort_values(["state", "month"])
g = st.groupby("state")["change"]
st["r1"]  = st["change"]
st["r3"]  = g.transform(lambda s: s.rolling(3).sum())
st["r6"]  = g.transform(lambda s: s.rolling(6).sum())
st["r12"] = g.transform(lambda s: s.rolling(12).sum())
st["future3"] = g.transform(lambda s: s[::-1].rolling(3).sum()[::-1].shift(-1))   # months t+1..t+3

ext = pd.DataFrame({"nat_r3": nat["change"].rolling(3).sum(),
                    "fx_r3": np.log(drivers["ngn_per_usd"]).diff(3)})
# (petrol is not used: it was not recorded Feb–Dec 2025 and added nothing in testing)
st = st.merge(ext, left_on="month", right_index=True, how="left")
st["msin"] = np.sin(2 * np.pi * st["month"].dt.month / 12)
st["mcos"] = np.cos(2 * np.pi * st["month"].dt.month / 12)

FEATS = ["r1", "r3", "r6", "r12", "nat_r3", "fx_r3", "msin", "mcos"]
panel = st.dropna(subset=FEATS + ["future3"]).copy()
panel["surge"] = (panel["future3"] > SURGE).astype(int)
train, test = panel[panel["month"] < SPLIT], panel[panel["month"] >= SPLIT]
print(f"Train {train['month'].min():%b %Y}–{train['month'].max():%b %Y}: {len(train)} state-months, "
      f"{100 * train['surge'].mean():.0f}% surges")
print(f"Test  {test['month'].min():%b %Y}–{test['month'].max():%b %Y}: {len(test)} state-months, "
      f"{100 * test['surge'].mean():.0f}% surges\n")

models = {
    "Logistic regression": make_pipeline(StandardScaler(), LogisticRegression(class_weight="balanced", max_iter=1000)),
    "Gradient boosting":   GradientBoostingClassifier(n_estimators=200, max_depth=2, learning_rate=0.05,
                                                      subsample=0.8, random_state=0),
}
scores = {"Rule: last 3 months rose > 10%": test["r3"].values}
for name, m in models.items():
    m.fit(train[FEATS], train["surge"])
    scores[name] = m.predict_proba(test[FEATS])[:, 1]

rows = []
for name, s in scores.items():
    thr = SURGE if name.startswith("Rule") else 0.5
    pred = (s > thr).astype(int)
    rows.append(dict(model=name, AUC=roc_auc_score(test["surge"], s),
                     caught_pct=100 * recall_score(test["surge"], pred),
                     alerts_correct_pct=100 * precision_score(test["surge"], pred, zero_division=0)))
ews_score = pd.DataFrame(rows).set_index("model").round(2)
print(ews_score.to_string())
print("\nAUC: 0.5 = coin flip, 1.0 = perfect.  caught = surges warned about;  alerts_correct = warnings that came true.")

lr = models["Logistic regression"]
coef = pd.Series(lr[-1].coef_[0], index=FEATS).sort_values(key=abs, ascending=False).round(2)
print("\nWhat raises the risk (standardised logistic coefficients):")
print(coef.to_string())

# Current risk: latest month with complete features for each state
cur = st.dropna(subset=FEATS).sort_values("month").groupby("state").tail(1)
cur = cur[cur["month"] >= LATEST - pd.DateOffset(months=6)].copy()
cur["risk_pct"] = 100 * lr.predict_proba(cur[FEATS])[:, 1]
cur = cur.sort_values("risk_pct", ascending=False)
print("\nCurrent surge risk (states still in the survey):")
print(cur[["state", "month", "risk_pct"]].assign(month=cur["month"].dt.strftime("%b %Y")).round(0).to_string(index=False))

fig = make_subplots(rows=1, cols=2, horizontal_spacing=0.12,
                    subplot_titles=["Test period: warning vs baseline (ROC)",
                                    "Current risk of a >10% rise in the next 3 months"])
for i, (name, s) in enumerate(scores.items()):
    fpr, tpr, _ = roc_curve(test["surge"], s)
    fig.add_trace(go.Scatter(x=fpr, y=tpr, name=f"{name} (score {ews_score.loc[name, 'AUC']:.2f})",
                             legendgroup="roc", legendgrouptitle_text="Left: warning method (1.0 = perfect)",
                             line=dict(width=2.3, color=["#9E9E9E", "#1565C0", "#EF6C00"][i])), row=1, col=1)
fig.add_trace(go.Scatter(x=[0, 1], y=[0, 1], name="Pure guessing (score 0.50)", legendgroup="roc",
                         line=dict(color="grey", dash="dot")), row=1, col=1)
for label, colour, lo, hi in [("Low risk (under 30 %)", "#2E7D32", -1, 30),
                              ("Medium risk (30–60 %)", "#EF6C00", 30, 60),
                              ("High risk (60 % or more)", "#C62828", 60, 101)]:
    part = cur[(cur["risk_pct"] >= lo) & (cur["risk_pct"] < hi)]
    fig.add_trace(go.Bar(x=part["state"], y=part["risk_pct"], marker_color=colour, name=label,
                         legendgroup="risk", legendgrouptitle_text="Right: current risk level",
                         text=[f"{r:.0f}%<br>({m:%b %Y})" for r, m in zip(part["risk_pct"], part["month"])],
                         textposition="outside"), row=1, col=2)
    if part.empty:                                   # keep the colour in the legend even when unused
        legend_item(fig, label, colour, legendgroup="risk", row=1, col=2)
fig.update_xaxes(title_text="Share of false alarms", row=1, col=1)
fig.update_yaxes(title_text="Share of real surges caught", row=1, col=1)
fig.update_xaxes(categoryorder="array", categoryarray=cur["state"].tolist(), row=1, col=2)
fig.update_yaxes(title_text="Risk (%)", range=[0, 110], row=1, col=2)
fig.update_layout(height=480, barmode="relative", legend=dict(groupclick="toggleitem"),
                  title="<b>10 · Early-warning system for food price surges</b>"
                        f"<br><sup>Trained on data before {SPLIT:%b %Y}, tested on {SPLIT:%b %Y} onwards.</sup>")
show(fig, "10_early_warning",
     "Left: how well each method warned about price surges (a rise of more than 10% within 3 months) in 2021–2025. "
     "A line hugging the<br>top-left corner catches most surges with few false alarms; the dotted diagonal is pure "
     "guessing; a higher score is better.<br>Right: each state's current chance of a surge in the next 3 months, "
     "coloured green / orange / red by risk level.")

Train Jul 2015–Dec 2020: 209 state-months, 38% surges
Test  Jan 2021–Nov 2025: 331 state-months, 26% surges



                                 AUC  caught_pct  alerts_correct_pct
model                                                               
Rule: last 3 months rose > 10%  0.70       60.92               46.90
Logistic regression             0.87       86.21               44.91
Gradient boosting               0.78       78.16               39.31

AUC: 0.5 = coin flip, 1.0 = perfect.  caught = surges warned about;  alerts_correct = warnings that came true.

What raises the risk (standardised logistic coefficients):
msin      2.20
mcos      1.20
r6        1.18
r3       -1.16
nat_r3    1.13
r12       0.75
fx_r3     0.73
r1        0.21

Current surge risk (states still in the survey):
state    month  risk_pct
Borno Feb 2026      28.0
 Yobe Feb 2026      16.0


### What we found
- A **"surge"** is defined as a state's food prices rising **more than 10 % over the next 3 months**. The model was
  trained on 2015–2020 and tested on Jan 2021 – Nov 2025 (331 state-months, 26 % of them surges).
- **The logistic-regression warning model scores AUC 0.87.** It caught **86 % of surges** before they happened, and
  **45 % of its warnings came true** (about 1 in 2).
- The simple rule "prices already rose 10 % in the last 3 months" only reaches AUC 0.70 and catches 61 % of surges.
  Gradient boosting reached 0.78; the simpler model generalises better.
- **Strongest warning signs:** the time of year (February–April, before the lean season), a steady rise over the last
  6 months, rising prices nationally, and the naira weakening in the last 3 months. A sudden 3-month spike on its own
  tends to partly reverse.
- **Current risk (Feb 2026): Borno 28 %, Yobe 16 %.** These are low to moderate.

### What it means for the man in the street
The warning system catches most price surges one to three months before they arrive. About half its alarms are
false, which is acceptable for "get ready" actions: stock grain, pre-position food aid, plan cash transfers. It is not
reliable enough to act on in a panic. **Right now there is no red flag for Borno or Yobe**, but the risk normally
rises as the lean season approaches.

---
## Summary — what all of this means for ordinary Nigerians

| # | Finding | What to do |
|---|---|---|
| 1 | **Food costs about 4.4× what it did in January 2019** (₦1,000 → ₦4,400). 2024 was the worst year: prices doubled. | Budget for food at today's prices, not 2023's. |
| 2 | **2025 brought real relief for grains and beans (−25 %)**, but **meat, fish, eggs and cooking oil are still rising**. | Protect children's protein: beans, groundnuts and eggs are the cheapest protein sources per naira. |
| 3 | **The naira drives food prices.** Measured in dollars, food costs the same as in 2019. | When the naira slides, stock up on dry goods within weeks. |
| 4 | **Prices follow the harvest calendar**: cheapest Oct–Dec, dearest Jun–Aug (20–25 % swing for grains). | Buy grain in bulk after harvest; farmers who can store safely should sell in the lean season. |
| 5 | **Processed food rose much more than raw food** (maize flour ×6.4 vs millet grain ×4.5). **Gari by the cup costs ~50 % more than by the bag.** | Buy whole grain and mill it; buy gari by the bag or share a bag. |
| 6 | **Lagos pays ~30 % more than the North**, but northern prices are rising faster. | Northern rural households need the most support. |
| 7 | **Outlook: about +14 % by August 2026** (range −5 % to +44 %). | Buy staples early; do not wait for the lean season. |

**Limits of this analysis.** The data come from WFP monitoring focused on the North-East. Since 2023 only Borno and Yobe
are covered, and most other states had just one market. This is not the official NBS CPI. Correlations show that
things move together; they do not prove cause. Forecasts beyond three months are uncertain.